[WIP] Pf remove goroutine

wojtek-t opened the pull request:
```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

wojtek-t committed:
Remove unused promise code from APF

wojtek-t committed:
Simplify APF promise to what is really used in the code

wojtek-t committed:
Eliminate goroutine

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:
```diff
@@ -39,20 +40,32 @@ type promise struct {
 		p.activeCounter.Add(-1)
-		p.cond.Wait()
+
+		lock.Unlock()
```
I don't think the locking pattern is what we really want - but that was supposed to illustrate the feasibility.

fedebongio commented:
/triage accepted

wojtek-t committed:
Cleanup locking model

MikeSpreitzer commented:
@MikeSpreitzer

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:
```diff
@@ -0,0 +1,140 @@
+	if !atomic.CompareAndSwapInt32(&p.isSet, 0, 1) {
+		return false
+	}
+	p.value = value
```
There is a timing splinter here.  Suppose some other goroutine calls `WaitAndGet` after `p.isSet` becomes 1 but before `p.value` is written.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:
```diff
@@ -0,0 +1,140 @@
+	p.value = value
+	close(p.doneCh)
+
+	p.unblockWaiting()
```
The goroutine counter has to be updated before the waiting goroutines are unblocked (in this case by the `close(p.doneCh)`); doing it later creates timing splinters.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:
```diff
@@ -0,0 +1,140 @@
+		// There is a possible race that between checking p.isSet and increasing
+		// p.waitingCount p.Set was called and did the waitingCount decrease.
+		// To protect against this, we need to check if we shouldn't decrease here too.
+		defer p.unblockWaiting()
```
There is a counting problem here.  If there are multiple goroutines in `WaitAndGet` and only a subset are unblocked via `stopCh` then the current logic incorrectly counts all of the waiters as re-activated.

MikeSpreitzer reviewed (commented):
I see three key ideas here.

1. Implement the promise with sync/atomic rather than mutex.  That looks problematic, I found multiple timing splinters.
2. Use different promise implementations in regular vs test code, with the goroutine counting stuff just in the test implementation.  That looks viable and not necessarily a very big change.
3. Making the testing promise wait on _either_ of two things, one of which is not counted.  That does big violence to the goroutine counting.  Easiest remedy is to keep the test code using the existing pattern of (a) promise waits on just one thing and (b) add a goroutine that waits on context cancel and then fulfills the promise.  A bigger and more fundamental change would be to revise the goroutine counting framework so that it counts channels and knows about union waits.  That would be significantly more complex.  Remember that we have an easy case here because we know for sure that each of the contexts in question will be canceled reasonably promptly and we know that the code in question (StartRequest), if using the first remedy in testing code, has exactly 1 goroutine that is activated by that cancel.  So I am drawn to the first remedy.

wojtek-t commented:
@MikeSpreitzer 

> Implement the promise with sync/atomic rather than mutex. That looks problematic, I found multiple timing splinters.

I'm not attached to the idea of using atomic (i was just experimenting with it).
I just really want to simplify the locking scheme and get rid of the goroutine which doesn't seem needed.

> Making the testing promise wait on either of two things, one of which is not counted. That does big violence to the goroutine counting. Easiest remedy is to keep the test code using the existing pattern of (a) promise waits on just one thing and (b) add a goroutine that waits on context cancel and then fulfills the promise. A bigger and more fundamental change would be to revise the goroutine counting framework so that it counts channels and knows about union waits. That would be significantly more complex. Remember that we have an easy case here because we know for sure that each of the contexts in question will be canceled reasonably promptly and we know that the code in question (StartRequest), if using the first remedy in testing code, has exactly 1 goroutine that is activated by that cancel. So I am drawn to the first remedy.

Allocating goroutines is non-negliigble overhead in large clusters where we may potentially have many of them (for every in-flight request either being processed or waiting in the queue). So I would really like to eliminate it.

I think I'm not following the argument about fundamental change here. The whole point of couting gorotuines currently is that we ensure that it may not do anything *relevant* if it's not being counted. This implementation (modulo the problems related to using atomic) doesn't really validate it - it's counting the goroutines the same way as it was before.

wojtek-t commented:
@MikeSpreitzer - to not introduce confusion, I removed the last commit and I switched to lock-based implementation.
I also cleaned up goroutine counting to make it self-contained in the file.

With this change - I believe that no further changed are needed anywhere and it eliminates the need for any goroutine.

PTAL

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:
```diff
@@ -0,0 +1,112 @@
+	select {
+	case <-p.doneCh:
+		return p.value, nil
+	case <-stopCh:
+		p.lock.Lock()
+		defer p.lock.Unlock()
+		// Check if in the meantime value wasn't set (and thus the goroutine wasn't activated).
+		select {
+		case <-p.doneCh:
+			return p.value, nil
+		default:
+		}
+		// Now we know we really timed out. So we reactivate the goroutine.
+		p.waitingCount--
```
For this range of code, there a goroutine running it and there is no blockage in sight, but the goroutine is not counted in the GoRoutineCounter.  This goroutine will be incorrectly ignored by a concurrent goroutine in FakeEventClock::Run.  For counting without timing splinters, the pattern that needs to be followed is to increase the count _before_ unblocking some existing goroutine or starting a new one.

wojtek-t replied:
And I do that - this is happening only in the case where I actually timeout, in which case there is no other way to count that [and this is compatible with how counting currently works]

MikeSpreitzer replied:
I suppose that by "and I do that" you mean https://github.com/kubernetes/kubernetes/blob/8bf5872d8da305b3987bd8344914769acb249d35/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go#L699-L700 .  That makes https://github.com/kubernetes/kubernetes/blob/8bf5872d8da305b3987bd8344914769acb249d35/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go#L86 double-counting of the same goroutine.  One of those two has to go.

If we delete the former then the timing splinter I described is present.  If we delete the latter then the code is only correct if every closure of `Context.Done()` is preceded by a counter increment _and_ we know that in that case the promise does not resume until that closure.  This is true in the one test case that exercises closure currently, but it seems an unnecessarily fragile pattern to establish.  I would rather use the more robust pattern with a goroutine in the test code.  I appreciate the desire to remove unnecessary costs from the APF filter in the non-test code.  What I am proposing in #104002 is to use a goroutine only in the test code.

MikeSpreitzer reviewed (commented):
I do not see a way to do correct goroutine counting in the normal code unless we do a pervasive replace of using channels with something non-channel-based or wrap all the channels (including the ones involved in a Context) with something that does counting.  Counting correctly has to account for the facts that (a) code closing a channel generally does not know how many goroutines wait on that channel and (b) a select can be on multiple channels.

The above is a much heavier lift than I would advocate.  Instead I originally took the approach of making the QueueSet code (normal and testing) do correct counting and not worry about incorrectness of the counting when used in non-testing contexts.  And show how to do non-channel-based synchronization so that usage of that could grow incrementally if people like it.

@wojtek-t you seem to be on a campaign to expunge the counting from the non-testing code.  I can live with that.  The remaining question is how to correctly handle the union-wait (select with multiple cases) in the test code.  I created the promise abstraction to wrap up the union-wait in a way that can be implemented with correct counting.  As your previous revision of this PR demonstrated, we could use a different promise implementation in the non-testing vs in the testing code.  As you see in #104002, the non-testing promise implementation can be rather simple.  So the remaining question is how to implement the union-wait correctly in the testing code.  I now see the following possibilities for the testing promise.

1. Taking advantage of the restricted usage (that is, knowledge that the stopCh is closed promptly), fork a goroutine that reacts to the stopCh closure with a call into the promise abstraction.  This is the approach I took in #104002, because it is the smallest change.
2. Let the testing promise select on multiple cases but also have logic to correctly count and handle this.  Because the `GoRoutineCounter::Add(1)` has to precede unblocking the select, this means that the testing promise has to have some code that runs just before closing the stopCh.  That, in turn, requires the queueSet test code to reach into the `queueset/request` to dig out the testing promise and invoke that extra code.  And it requires some non-trivial expansion of the promise data structure and code to correctly handle the union-waits.  This is an intermediate possibility that I did not mention in my earlier comment.
3. Handle the union-waits more fundamentally in the FakeEventClock rather than the testing promise implementation.  This is what I tried to outline in my earlier comment.

wojtek-t commented:
@MikeSpreitzer - I looked a bit deeper into it and then evolution I wanted to do is actually aligned with what your PR does.

So I'm closing this in favor of #104002  - let's proceed with that one.
